modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,568 @@
+    let mut check_ins = DAILY_CHECK_INS.load(deps.storage, challenge_id.clone())?;
+    let check_in = check_ins.last_mut().unwrap();
+
+    if check_in.status != CheckInStatus::CheckedInNotYetVoted {
+        return Err(AppError::WrongCheckInStatus {});
+    }
+
+    // Check if the voter has already voted
+    if VOTES
+        .may_load(
+            deps.storage,
+            (challenge_id.to_owned(), vote.voter.to_owned()),
+        )
+        .map_or(false, |votes| votes.iter().any(|v| v.voter == vote.voter))
+    {
+        return Err(AppError::AlreadyVoted {});
+    }
+
+    VOTES.save(
+        deps.storage,
+        (challenge_id.to_owned(), vote.voter.to_owned()),
+        &vote,
+    )?;
+    let mut challenge_votes = CHALLENGE_VOTES.load(deps.storage, challenge_id.clone())?;
+    vote.for_check_in = Some(check_in.last_checked_in);
+    challenge_votes.push(vote);
+    CHALLENGE_VOTES.save(deps.storage, challenge_id.clone(), &challenge_votes)?;
+
+    check_in.status = CheckInStatus::VotedNotYetTallied;
+    DAILY_CHECK_INS.save(deps.storage, challenge_id, &check_ins)?;
+
+    // If all friends have voted, we tally the votes.
+    if CHALLENGE_VOTES
+        .load(deps.storage, challenge_id.clone())?
+        .len()
+        == CHALLENGE_FRIENDS.load(deps.storage, challenge_id)?.len()
+    {
+        tally_votes_for_check_in(deps, env, challenge_id)?;
+    }
+    Ok(Response::new().add_attribute("action", "cast_vote"))
+}
+
+fn tally_votes_for_check_in(deps: DepsMut, _env: Env, challenge_id: u64) -> AppResult {
+    let mut challenge = CHALLENGE_LIST.load(deps.storage, challenge_id.clone())?;
+    // if challenge.status != ChallengeStatus::OverAndPending {
+    //     return Err(AppError::WrongChallengeStatus {});
+    // }
+
+    let check_ins = DAILY_CHECK_INS.load(deps.storage, challenge_id.clone())?;
+    let check_in = check_ins.last().unwrap();
+
+    if check_in.status != CheckInStatus::VotedNotYetTallied
+        || check_in.status != CheckInStatus::Recount
+    {
+        return Err(AppError::WrongCheckInStatus {});
+    }
+
+    let votes = CHALLENGE_VOTES.load(deps.storage, challenge_id)?;
+
+    // check for any false votes on check_ins that match the vote timestamps
+    let any_false_vote = votes
+        .iter()
+        .filter(|&vote| {
+            vote.for_check_in
+                .map_or(false, |timestamp| timestamp == check_in.last_checked_in)
+        })
+        .any(|v| v.approval == Some(false));
+
+    //@TODO only update ChallengeStatus if the challenge is OverAndPending
+    // update check_in status
+    // update check_in tally_result
+    //
+    // Then charge penalty and there check if the check_in tally_result is Some(false)
+    // if yes divide the collateral among how many days there are between now
+    // and challenge.end and split that amount between friends.
+
+    if any_false_vote {
+        challenge.status = ChallengeStatus::OverAndFailed;
```

Comment:
So the challenge is failed if one day's check-in is failed?

Reply:
this wasn't fully implemented when you reviewed, it's been updated now!